File: src/main/java/com/github/kindrat/cassandra/client/filter/DataFilter.java
package com.github.kindrat.cassandra.client.filter;

import com.datastax.driver.core.Row;
import com.github.kindrat.cassandra.client.filter.fields.*;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.StringUtils;

import java.util.EnumMap;
import java.util.Map;
import java.util.Optional;
import java.util.function.Predicate;

import static com.github.kindrat.cassandra.client.filter.Combiner.AND;
import static com.github.kindrat.cassandra.client.filter.Combiner.OR;
import static com.github.kindrat.cassandra.client.util.StringUtil.findClosingBracketIndex;
import static com.github.kindrat.cassandra.client.util.StringUtil.wordAtPosition;
import static com.github.kindrat.cassandra.client.util.UIUtil.parseWords;
import static org.apache.commons.lang3.StringUtils.startsWithIgnoreCase;

@Slf4j
public class DataFilter {
    private static Map<Operator, PredicateProvider> predicateFactories = new EnumMap<>(Operator.class);

    static {
        predicateFactories.put(Operator.EQ, EqPredicate::new);
        predicateFactories.put(Operator.GT, GtPredicate::new);
        predicateFactories.put(Operator.GE, GePredicate::new);
        predicateFactories.put(Operator.LT, LtPredicate::new);
        predicateFactories.put(Operator.LE, LePredicate::new);
        predicateFactories.put(Operator.NE, NePredicate::new);
        predicateFactories.put(Operator.LIKE, LikePredicate::new);
    }

    public static Predicate<Row> parsePredicate(String filter) {
        if (filter.startsWith("(")) {
            int closingBracketIndex = findClosingBracketIndex(filter);
            Predicate<Row> predicate = parsePredicate(filter.substring(1, closingBracketIndex));

            if (closingBracketIndex != filter.length() - 1) {
                String secondFilterPart = filter.substring(closingBracketIndex + 1).trim();
                if (startsWithIgnoreCase(secondFilterPart, AND.name())) {
                    return predicate.and(parsePredicate(secondFilterPart.substring(3).trim()));
                } else if (startsWithIgnoreCase(secondFilterPart, OR.name())) {
                    return predicate.or(parsePredicate(secondFilterPart.substring(2).trim()));
                } else {
                    throw new IllegalArgumentException("Illegal combiner : " + secondFilterPart);
                }
            } else {
                return predicate;
            }
        }
        String[] words = parseWords(filter);

        if (words.length < 3) {
            throw new IllegalArgumentException("Incomplete filter : " + filter);
        }
        String firstWord = wordAtPosition(words, 0)
                .orElseThrow(() -> new IllegalArgumentException("Should not be called with empty string"));
        String secondWord = wordAtPosition(words, 1)
                .orElseThrow(() -> new IllegalArgumentException("Should not be called with empty string"));
        String thirdWord = wordAtPosition(words, 2)
                .orElseThrow(() -> new IllegalArgumentException("Should not be called with empty string"));

        Operator operator = Operator.valueOf(secondWord);
        Predicate<Row> operatorPredicate = predicateFactories.get(operator).build(firstWord, thirdWord);

        Optional<String> fourthWord = wordAtPosition(words, 3);
        if (fourthWord.isPresent()) {
            Combiner combiner = Combiner.valueOf(fourthWord.get());
            String[] remainingWords = ArrayUtils.subarray(words, 4, words.length);
            String remainingFilterString = StringUtils.join(remainingWords, " ");
            switch (combiner) {
                case AND:
                    return operatorPredicate.and(parsePredicate(remainingFilterString));
                case OR:
                    return operatorPredicate.or(parsePredicate(remainingFilterString));
            }
        }

        return operatorPredicate;
    }

    @FunctionalInterface
    private interface PredicateProvider {
        RowPredicate build(String field, String value);
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/filter/DataFilter.kt
package com.github.kindrat.cassandra.client.filter

import com.github.kindrat.cassandra.client.filter.fields.*
import com.github.kindrat.cassandra.client.ui.DataObject
import com.github.kindrat.cassandra.client.util.StringUtil
import com.github.kindrat.cassandra.client.util.UIUtil
import java.util.*
import java.util.function.Predicate

object DataFilter {
    private val predicateFactories: MutableMap<Operator, PredicateProvider> = EnumMap(Operator::class.java)

    @JvmStatic
    fun parsePredicate(filter: String): Predicate<DataObject> {
        if (filter.startsWith("(")) {
            val closingBracketIndex = StringUtil.findClosingBracketIndex(filter)
            val predicate = parsePredicate(filter.substring(1, closingBracketIndex))
            return if (closingBracketIndex != filter.length - 1) {
                val secondFilterPart = filter.substring(closingBracketIndex + 1).trim { it <= ' ' }
                when {
                    secondFilterPart.startsWith(Combiner.AND.name, ignoreCase = true) -> {
                        predicate.and(parsePredicate(secondFilterPart.substring(3).trim { it <= ' ' }))
                    }

                    secondFilterPart.startsWith(Combiner.OR.name, ignoreCase = true) -> {
                        predicate.or(parsePredicate(secondFilterPart.substring(2).trim { it <= ' ' }))
                    }
                    else -> {
                        throw IllegalArgumentException("Illegal combiner : $secondFilterPart")
                    }
                }
            } else {
                return predicate
            }
        }
        val words = UIUtil.parseWords(filter)
        require(words.size >= 3) { "Incomplete filter : $filter" }
        val firstWord = StringUtil.wordAtPosition(words, 0) ?:
            throw IllegalArgumentException("Should not be called with empty string")
        val secondWord = StringUtil.wordAtPosition(words, 1) ?:
            throw IllegalArgumentException("Should not be called with empty string")
        val thirdWord = StringUtil.wordAtPosition(words, 2) ?:
            throw IllegalArgumentException("Should not be called with empty string")
        val operator: Operator = Operator.fromValue(secondWord)
        val operatorPredicate: Predicate<DataObject> = predicateFactories[operator]!!.build(firstWord, thirdWord)
        val fourthWord = StringUtil.wordAtPosition(words, 3)
        fourthWord?.let {
            val combiner = Combiner.valueOf(it)
            val remainingWords = words.copyOfRange(4, words.size)
            val remainingFilterString = remainingWords.joinToString { " " }
            return when (combiner) {
                Combiner.AND -> operatorPredicate.and(parsePredicate(remainingFilterString))
                Combiner.OR -> operatorPredicate.or(parsePredicate(remainingFilterString))
            }
        }
        return operatorPredicate
    }

    fun interface PredicateProvider {
        fun build(field: String, value: String): RowPredicate
    }

    init {
        predicateFactories[Operator.EQ] = PredicateProvider { field, value -> EqPredicate(field, value) }
        predicateFactories[Operator.GT] = PredicateProvider { field, value -> GtPredicate(field, value) }
        predicateFactories[Operator.GE] = PredicateProvider { field, value -> GePredicate(field, value) }
        predicateFactories[Operator.LT] = PredicateProvider { field, value -> LtPredicate(field, value) }
        predicateFactories[Operator.LT] = PredicateProvider { field, value -> LePredicate(field, value) }
        predicateFactories[Operator.NE] = PredicateProvider { field, value -> NePredicate(field, value) }
        predicateFactories[Operator.LIKE] = PredicateProvider { field, value -> LikePredicate(field, value) }
    }
}